hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/replicas/ReplicaStateVerifier.java

```diff
@@ -0,0 +1,222 @@
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Objects;
+import org.apache.hadoop.hdds.client.StandaloneReplicationConfig;
+import org.apache.hadoop.hdds.conf.OzoneConfiguration;
+import org.apache.hadoop.hdds.protocol.DatanodeDetails;
+import org.apache.hadoop.hdds.protocol.datanode.proto.ContainerProtos.ContainerDataProto;
+import org.apache.hadoop.hdds.protocol.datanode.proto.ContainerProtos.ReadContainerResponseProto;
+import org.apache.hadoop.hdds.scm.XceiverClientManager;
+import org.apache.hadoop.hdds.scm.XceiverClientSpi;
+import org.apache.hadoop.hdds.scm.cli.ContainerOperationClient;
+import org.apache.hadoop.hdds.scm.container.ContainerInfo;
+import org.apache.hadoop.hdds.scm.pipeline.Pipeline;
+import org.apache.hadoop.hdds.scm.storage.ContainerProtocolCalls;
+import org.apache.hadoop.ozone.om.helpers.OmKeyLocationInfo;
+
+/**
+ * Verifies the state of a replica from the DN.
+ * [DELETED, UNHEALTHY, INVALID] are considered bad states.
+ */
+public class ReplicaStateVerifier implements ReplicaVerifier {
+  private static final String CHECK_TYPE = "replicaState";
+  private final ContainerOperationClient containerOperationClient;
+  private final XceiverClientManager xceiverClientManager;
+  // cache for replica details from the DNs
+  private final Cache<ReplicaKey, ContainerDataProto> containerDataCache = CacheBuilder.newBuilder()
+      .maximumSize(1000)
+      .build();
+  // cache for container info and encodedToken from the SCM
+  private final Cache<Long, ContainerInfoToken> encodedTokenCache = CacheBuilder.newBuilder()
+      .maximumSize(1000)
+      .build();
+
+  public ReplicaStateVerifier(OzoneConfiguration conf) throws IOException {
+    containerOperationClient = new ContainerOperationClient(conf);
+    xceiverClientManager = containerOperationClient.getXceiverClientManager();
+  }
+
+  @Override
+  public String getType() {
+    return CHECK_TYPE;
+  }
+
+  @Override
+  public BlockVerificationResult verifyBlock(DatanodeDetails datanode, OmKeyLocationInfo keyLocation,
+                                             int replicaIndex) {
+    try {
+      StringBuilder replicaCheckMsg = new StringBuilder().append("Replica state is ");
+      boolean pass = false;
+
+      ContainerInfoToken containerInfoToken = getContainerInfoToken(keyLocation.getContainerID());
+      ContainerInfo containerInfo = containerInfoToken.getContainerInfo();
+
+      ContainerDataProto containerData = getContainerData(datanode, keyLocation, replicaIndex, containerInfoToken);
+      ContainerDataProto.State state = containerData.getState();
+      if (state == ContainerDataProto.State.UNHEALTHY ||
+          state == ContainerDataProto.State.INVALID ||
+          state == ContainerDataProto.State.DELETED) {
+        replicaCheckMsg.append(state.name());
+      } else {
+        pass = true;
+      }
+      replicaCheckMsg.append(", Container state in SCM is ").append(containerInfo.getState());
```

Comment:
We should check if the state in SCM is DELETING or DELETED first, and that should also fail the test. I would still check the replicas after this regardless of SCM state, just to see if there's weird cases where the container is marked deleted but replicas still exist.